### PR TITLE
feat: add filterFourierTransformArtefacts function (issue #223)

### PR DESCRIPTION
**Pull request by jorainer**
This PR adds the code from @stanstrup to remove Orbitrap shoulders/fourier transform artefacts from spectra.

### CODE REVIEW

**Comment by jorainer** [comment]
I've implemented a modified version of the function that performs better on larger spectra:

```r
.peaks_remove_fft_artifact2 <- function(x, halfWindowSize = 0.05,
                                        threshold = 0.2,
                                        keepIsotopes = TRUE,
                                        maxCharge = 5,
                                        isotopeTolerance = 0.005) {
    neutron   <- 1.0033548378 # really C12, C13 difference
    iso_dist  <- neutron / seq(from = 1, by = 1, to = maxCharge)
    find_isotopes <- keepIsotopes & any(iso_dist < halfWindowSize)
    mz <- x[, "mz"]
    int <- x[, "intensity"]
    idx <- order(int, decreasing = TRUE)
    keep <- rep(TRUE, length(idx))
    while (!is.na(i <- idx[1L])) {
        idx <- idx[-1L]
        rem_candidate <- which(
            between(mz[idx], mz[i] + c(-halfWindowSize, halfWindowSize)))
        rem_candidate <- rem_candidate[int[idx][rem_candidate] / int[i] <
                                       threshold]
        ## That part might be improved later.
        if (find_isotopes) {
            target_dist <- abs(mz[idx][rem_candidate] - mz[i])
            dist_matrix <- outer(target_dist, iso_dist, "-")
            dist_matrix <- abs(dist_matrix)
            dist_to_iso_hypo <- apply(dist_matrix, 1, min)
            rem_candidate <- rem_candidate[dist_to_iso_hypo > isotopeTolerance]
        }
        if (length(rem_candidate)) {
            keep[idx[rem_candidate]] <- FALSE
            idx <- idx[-rem_candidate]
        }
    }
    x[keep, , drop = FALSE]
}
```

This function considers in each loop only peaks that were not already removed or also tested - thus, it gets faster with each iteration:

```r
data(fft_spectrum)
library(microbenchmark)
sp <- peaksData(fft_spectrum)[[1L]]
microbenchmark(Spectra:::.peaks_remove_fft_artifact(sp),
               .peaks_remove_fft_artifact2(sp))
Unit: milliseconds
                                     expr    min      lq     mean  median
 Spectra:::.peaks_remove_fft_artifact(sp) 4.7659 5.57925 6.451996 5.80830
          .peaks_remove_fft_artifact2(sp) 2.7591 3.19880 3.920461 3.38825
      uq     max neval cld
 6.00455 16.5911   100   b
 3.53515 18.4121   100  a 

sciex_file <- normalizePath(
    dir(system.file("sciex", package = "msdata"), full.names = TRUE))
sciex_mzr <- backendInitialize(MsBackendMzR(), files = sciex_file)
sp <- peaksData(sciex_mzr)[[1705L]]
microbenchmark(Spectra:::.peaks_remove_fft_artifact(sp),
               .peaks_remove_fft_artifact2(sp))
Unit: milliseconds
                                     expr      min        lq      mean
 Spectra:::.peaks_remove_fft_artifact(sp) 193.9958 205.78610 223.92499
          .peaks_remove_fft_artifact2(sp)  43.3130  46.91455  49.58035
    median       uq      max neval cld
 213.67975 220.5217 483.4687   100   b
  48.39205  51.9478  61.1351   100  a 

```

**Comment by jorainer** [comment]
@stanstrup, can you please have a look at this PR to see if it's OK from your side?

**Comment by stanstrup** [comment]
Looks good :)
Wouldn't it make sense to set ylim so you clearly see the issue?

```
## Removing fourier transform artefacts seen in Orbitra data.

## Loading an Orbitrap spectrum with artefacts.
data(fft_spectrum)
length(mz(fft_spectrum)[[1]])
plotSpectra(fft_spectrum, xlim = c(264.5, 265.5))

plotSpectra(fft_spectrum, xlim = c(264.5, 265.5), ylim = c(0, 5e6))


fft_spectrum_filtered <- filterFourierTransformArtefacts(fft_spectrum)

fft_spectrum_filtered
length(mz(fft_spectrum_filtered)[[1]])
plotSpectra(fft_spectrum_filtered, xlim = c(264.5, 265.5), ylim = c(0, 5e6))


# Using a few examples peaks in your data you can optimize the parameters
fft_spectrum_filtered <- filterFourierTransformArtefacts(fft_spectrum,
                                                halfWindowSize = 0.2,
                                                  threshold = 0.005,
                                                  keepIsotopes = TRUE,
                                                  maxCharge = 5,
                                                  isotopeTolerance = 0.005
                                                  )

fft_spectrum_filtered
length(mz(fft_spectrum_filtered)[[1]])
plotSpectra(fft_spectrum_filtered, xlim = c(264.5, 265.5), ylim = c(0, 5e6))


```

**Comment by stanstrup** [comment]
Something seems wrong in the logging:
`Remove fast fourier artefacts within. [Tue Nov 23 10:47:49 2021]`

**Comment by stanstrup** [comment]
Looks good.

File: R/MsBackendMzR-functions.R
Comment (lgatto) [inline R/MsBackendMzR-functions.R:30]: Quick question - when would `hdr` not be a `data.frame`? Shouldn't this rather be guaranteed by `mzR::header()`?

Comment (jorainer) [reply R/MsBackendMzR-functions.R:30]: good question - actually it's not. The *original* `mzR` code returns a `list` if the header of only one spectrum is extracted and a `data.frame` otherwise. We could this change in `mzR` (since it's also *cleaner*), but I'm afraid that might break code in other packages?

Comment (jorainer) [reply R/MsBackendMzR-functions.R]: I reverted that again. `mzR` did return before already a `data.frame` (otherwise this would have never worked before) - although documentation did tell something different. Based on the documentation I changed then the code in `mzR` in a recent PR (which never got into mzR *master* so far), which I did now again revert. So, to summarize, `mzR::header` did always return a `data.frame` and will return always a `data.frame`.

File: R/Spectra.R
Comment (lgatto) [inline R/Spectra.R:322]: lower than `threshold` times ~of~ the current peak's intensity

**Review by lgatto — COMMENTED** [review]
Looks good to me. I would nevertheless find it useful to read a bit more about that artefact, where it comes from, how to detect if, or a reference to read more about it.

**Comment by stanstrup** [comment]
> Looks good to me. I would nevertheless find it useful to read a bit more about that artefact, where it comes from, how to detect if, or a reference to read more about it.

I have never seen it described in the literature even if it is relatively well-known. What I know is here: https://github.com/sneumann/xcms/issues/94#issuecomment-336829325

**Comment by lgatto** [comment]
I think it's great if this is indeed explicitly described (in written, with code) for the first time in the software, which begs for a description of the artefact. That short paragraph is a great start! 

I don't know whether it should come under the details section of the Spectra man page, as part of the description of the function, or whether it warrants its own man page with details.

**Comment by jorainer** [comment]
That's an excellent idea. I think a dedicated, separate, man page for this function might be a good location for this information. So we have in the `Spectra` man page the description we have currently, but point from there to the more detailed man page.

Also, I think we can/should mention that this would be a first ad-hoc implementation and maybe more sophisticated versions will follow (if needed).

**Comment by jorainer** [comment]
I added a separate documentation file/man page with commit https://github.com/rformassspectrometry/Spectra/commit/93bae68d248c7073b8a865d6144136ea7350a832

**Review by lgatto — APPROVED** [review]

**Comment by lgatto** [comment]
Great! I also just realise that this new man page could possibly also be merge with the fft_spectrum manual.

**Comment by jorainer** [comment]
Makes sense - I'll combine both.

**Comment by jorainer** [comment]
Only problem we have now is that unit tests fail because of `mzR`.

File: R/peaks-functions.R
Comment (sgibb) [inline R/peaks-functions.R:458]: `MsCoreUtils::common` is needed for the suggested `find_isotopes` part. `MsCoreUtils::between` could removed if the suggested alternative implementation is accepted.

```suggestion
#' @importFrom MsCoreUtils common
```

File: R/peaks-functions.R
Comment (sgibb) [inline R/peaks-functions.R:491]: I like to suggest an alternative implementation here.
1. I would replace the `find_isotopes` part with `MsCoreUtils::common` (internally `MsCoreUtils::closests`).
2. I would take the index search outside the loop. In this context I would replace `MsCoreUtils::between` by two `findInterval` calls.

```suggestion
    ## just calculate isotopes that are in the halfWindowSize
    iso_dist <- iso_dist[iso_dist < halfWindowSize]
    find_isotopes <- keepIsotopes & length(iso_dist)
    if (find_isotopes)
        iso_dist <- c(-iso_dist, rev(iso_dist))

    mz <- x[, "mz"]
    int <- x[, "intensity"]

    ## left boundary
    lb <- findInterval(mz - halfWindowSize, mz) + 1L
    ## right boundary
    rb <- findInterval(mz + halfWindowSize, mz)

    ## region of interest (we just need to test if the window spans more than 1
    ## index)
    roi <- rb > lb

    ## test from the largest intensity
    idx <- seq_along(int)[roi][order(int[roi], decreasing = TRUE)]
    keep <- rep(TRUE, length(mz))

    for (i in idx) {
        if (keep[i]) {
            rem_candidate <- seq.int(lb[i], rb[i], by = 1L)
            rem_candidate <-
                rem_candidate[int[rem_candidate] / int[i] < threshold]

            if (find_isotopes) {
                cmm <- common(
                    mz[rem_candidate],
                    mz[i] + iso_dist,
                    tolerance = isotopeTolerance,
                    duplicates = "closest",
                    .check = FALSE
                )
                rem_candidate <- rem_candidate[!cmm]
            }

            keep[rem_candidate] <- FALSE
        }
    }
```

This changes yield a 5~6 runtime improvement in the aforementioned examples:
```r
library("Spectra")
library("MsCoreUtils")

.peaks_remove_fft_artifact2 <- function(x, halfWindowSize = 0.05,
                                        threshold = 0.2,
                                        keepIsotopes = TRUE,
                                        maxCharge = 5,
                                        isotopeTolerance = 0.005) {
    neutron   <- 1.0033548378 # really C12, C13 difference

    iso_dist <- neutron / seq.int(from = 1L, to = maxCharge, by = 1L)

    ## just calculate isotopes that are in the halfWindowSize
    iso_dist <- iso_dist[iso_dist < halfWindowSize]
    find_isotopes <- keepIsotopes & length(iso_dist)
    if (find_isotopes)
        iso_dist <- c(-iso_dist, rev(iso_dist))

    mz <- x[, "mz"]
    int <- x[, "intensity"]

    ## left boundary
    lb <- findInterval(mz - halfWindowSize, mz) + 1L
    ## right boundary
    rb <- findInterval(mz + halfWindowSize, mz)

    ## region of interest (we just need to test if the window spans more than 1
    ## index)
    roi <- rb > lb

    ## test from the largest intensity
    idx <- seq_along(int)[roi][order(int[roi], decreasing = TRUE)]
    keep <- rep(TRUE, length(mz))

    for (i in idx) {
        if (keep[i]) {
            rem_candidate <- seq.int(lb[i], rb[i], by = 1L)
            rem_candidate <-
                rem_candidate[int[rem_candidate] / int[i] < threshold]

            if (find_isotopes) {
                cmm <- common(
                    mz[rem_candidate],
                    mz[i] + iso_dist,
                    tolerance = isotopeTolerance,
                    duplicates = "closest",
                    .check = FALSE
                )
                rem_candidate <- rem_candidate[!cmm]
            }

            keep[rem_candidate] <- FALSE
        }
    }
    x[keep, , drop = FALSE]
}

data(fft_spectrum)
library(microbenchmark)
sp <- peaksData(fft_spectrum)[[1L]]
microbenchmark(
    Spectra:::.peaks_remove_fft_artifact(sp),
    .peaks_remove_fft_artifact2(sp),
    check = "identical"
)
# Unit: microseconds
#                                      expr      min        lq      mean
#  Spectra:::.peaks_remove_fft_artifact(sp) 2934.262 2972.7795 3527.2572
#           .peaks_remove_fft_artifact2(sp)  421.241  446.2135  510.7289
#   median        uq      max neval
#  3022.68 3170.3490 9368.680   100
#   527.62  561.3365  657.874   100

## with isotope checking
microbenchmark(
    Spectra:::.peaks_remove_fft_artifact(sp, halfWindowSize = 1),
    .peaks_remove_fft_artifact2(sp, halfWindowSize = 1),
    check = "identical"
)

# Unit: milliseconds
#                                                          expr      min
#  Spectra:::.peaks_remove_fft_artifact(sp, halfWindowSize = 1) 8.331218
#           .peaks_remove_fft_artifact2(sp, halfWindowSize = 1) 2.087970
#        lq     mean   median       uq       max neval
#  8.535107 9.114861 8.710721 8.873770 13.359157   100
#  2.226285 2.386233 2.294129 2.394645  7.202417   100

sciex_file <- normalizePath(
    dir(system.file("sciex", package = "msdata"), full.names = TRUE))
sciex_mzr <- backendInitialize(MsBackendMzR(), files = sciex_file)
sp <- peaksData(sciex_mzr)[[1705L]]
microbenchmark(
    Spectra:::.peaks_remove_fft_artifact(sp),
    .peaks_remove_fft_artifact2(sp),
    check = "identical"
)
# Unit: microseconds
#                                      expr      min        lq      mean
#  Spectra:::.peaks_remove_fft_artifact(sp) 2798.627 2886.6220 3355.8434
#           .peaks_remove_fft_artifact2(sp)  401.261  418.6355  487.2826
#     median       uq      max neval
#  2970.8425 3067.638 8554.893   100
#   467.0105  517.454 1283.004   100
```

File: R/peaks-functions.R
Comment (sgibb) [inline R/peaks-functions.R:466]: Would it be useful to make this an argument? (I don't know if anybody ever want's to change this.)

**Review by sgibb — CHANGES_REQUESTED** [review]
Sorry, I am a bit late to this PR. Thanks @jorainer and @stanstrup for this PR.
However I like to suggest an alternative implementation of the core function `.peaks_remove_fft_artifact`.
(everything else looks fine for me!)

File: R/peaks-functions.R
Comment (lgatto) [reply R/peaks-functions.R:491]: And again, @sgibb comes out of nowhere and between two nigh shifts proposes an alternative implementation that substantially speeds things up. So impressive!

Comment (stanstrup) [reply R/peaks-functions.R:491]: That is indeed damn impressive and a really meaningful improvement!

File: R/peaks-functions.R
Comment (stanstrup) [reply R/peaks-functions.R:466]: The only scenario I can think of where you want to change this is if you run an orbitrap in super duper high resolution mode so that there is isotopic fine-pattern. In that case though you'd need to define a vector of possible distances and not a single number...

File: R/peaks-functions.R
Comment (jorainer) [reply R/peaks-functions.R:491]: Great indeed @sgibb ! Only, I would change `duplicates = "closest"` to `duplicates = "keep"` - we've still some issues with the *closest* option (there is an issue is `MsCoreUtils`), and we also want to be a little bit more permissive keeping all possible isotope peaks.

File: R/peaks-functions.R
Comment (jorainer) [reply R/peaks-functions.R:466]: In the longer run we could have here also other isotopes (not only 13C, but also combinations etc). But for now I would keep it like this.

**Comment by jorainer** [comment]
@sgibb , I have now included all your suggestions. Please have a final look. Once you give me your OK I will merge.

**Pull request merged**
